analyzers/src/SonarAnalyzer.CSharp/Rules/ClassNotInstantiatable.cs

```diff
@@ -31,54 +33,33 @@ namespace SonarAnalyzer.Rules.CSharp
 
-        protected override bool IsObjectCreation(SyntaxNode node) =>
-            node.IsAnyKind(SyntaxKind.ObjectCreationExpression, SyntaxKindEx.ImplicitObjectCreationExpression);
-
-        private void CheckClassWithOnlyUnusedPrivateConstructors(SymbolAnalysisContext context)
+        protected override IEnumerable<Tuple<SyntaxNodeAndSemanticModel<BaseTypeDeclarationSyntax>, Diagnostic>> CollectRemovableDeclarations(INamedTypeSymbol namedType, Compilation compilation,
+            int count)
         {
-            var namedType = (INamedTypeSymbol)context.Symbol;
-            if (!IsNonStaticClassWithNoAttributes(namedType) || DerivesFromSafeHandle(namedType))
-            {
-                return;
-            }
-
-            var members = namedType.GetMembers();
-            var constructors = GetConstructors(members).Where(x => !x.IsImplicitlyDeclared).ToList();
+            var typeDeclarations = new CSharpRemovableDeclarationCollector(namedType, compilation).TypeDeclarations;
 
-            if (!HasOnlyCandidateConstructors(constructors) || HasOnlyStaticMembers(members.Except(constructors).ToList()))
-            {
-                return;
-            }
+            var message = count > 1
+                ? "at least one of its constructors"
+                : "its constructor";
 
-            var typeDeclarations = new CSharpRemovableDeclarationCollector(namedType, context.Compilation).TypeDeclarations;
-
-            if (!IsAnyConstructorCalled(namedType, typeDeclarations))
-            {
-                var message = constructors.Count > 1
-                    ? "at least one of its constructors"
-                    : "its constructor";
-
-                foreach (var typeDeclaration in typeDeclarations)
-                {
-                    var declarationKind = typeDeclaration.SyntaxNode.IsKind(SyntaxKind.ClassDeclaration)
-                        ? "class"
-                        : "record";
-                    context.ReportDiagnosticIfNonGenerated(
-                        Diagnostic.Create(Rule, typeDeclaration.SyntaxNode.Identifier.GetLocation(), declarationKind, message));
-                }
-            }
+            return typeDeclarations
+                .Select(x => new Tuple<SyntaxNodeAndSemanticModel<BaseTypeDeclarationSyntax>, Diagnostic>(x, Diagnostic.Create(rule,
+                                                                                                                               x.SyntaxNode.Identifier.GetLocation(),
+                                                                                                                               DeclarationKind(x.SyntaxNode),
+                                                                                                                               message)));
         }
+
+        private static string DeclarationKind(SyntaxNode node) =>
+            node.IsKind(SyntaxKind.ClassDeclaration)
```

Comment:
This is as short and simple that it can be single line. But I'd swap the logic to fall back to `class` in case something new and unexpected appears

Reply:
I also mentioned that it can be single line, because it's short and simple